### ISSUE TITLE
`compute_wtxid` does not consider coinbase inputs

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3313]
See discussion here: https://github.com/rust-bitcoin/rust-bitcoin/pull/3308#discussion_r1745771809

**Comment by tcharding** [comment]
Relevant code from the `block` module:

```rust
    /// Computes the Merkle root of transactions hashed for witness.
    pub fn witness_root(&self) -> Option<WitnessMerkleNode> {
        let hashes = self.txdata.iter().enumerate().map(|(i, t)| {
            if i == 0 {
                // Replace the first hash with zeroes.
                Wtxid::all_zeros()
            } else {
                t.compute_wtxid()
            }
        });
        WitnessMerkleNode::calculate_root(hashes)
    }
```

**Comment by Kixunil** [comment]
That code is correct, although likely not that efficient.

**Comment by apoelstra** [comment]
My proposal is to keep `compute_wtxid` as-is (not trying to detect coinbase inputs) but put a big warning on there. If we ever add a `cfg(debug_suspicious)` flag we can log if somebody tries to call this on a coinbase transaction.

This method is obscure enough, and calling it on a coinbase tx even more obscure, that I don't like the idea of splitting the function into two things with super long names (as Kix proposed in the other thread).

We can also look at the cost/benefit here:

* If somebody tries to call `compute_wtxid` to build a wtxid commitment root for their coinbase, thinking our method will handle the coinbase case when it doesn't, their result will be unambiguously wrong and harmlessly fail to verfiy. (Lol I guess maybe they could mine a block this way and waste some money, but I hope they'd test their code before putting big power into it.)
* If somebody tries to call it to populate a hashtable or something, and it **does** treat all coinbase-carrying inputs as the same but the user doesn't realize it, then entries in their table will collide and they'll have some potentially-subtle potentially-serious logic bug where e.g. somebody could silently replace coinbase transactions or something.

It's hard for me to think of a realistic situation where the latter case would occur, especially one where it even *matters* what happens with coinbase transactions. But one failure mode is super visible and the other less so, and the visible one is what we've already been doing ever since the wtxid method was introduced, so IMO we should stick with it.

>That code is correct, although likely not that efficient.

I run into this pattern a lot "do something special on the first iteration" and I'm never sure how to handle it best.

**Comment by Kixunil** [comment]
> with super long names

Maybe there are shorter alternatives, I just couldn't think of any so far.

Anyway, your analysis sounds good. wTXID is pretty obscure anyway, so I'm willing to accept it unless someone suggests something better.

> I run into this pattern a lot "do something special on the first iteration" and I'm never sure how to handle it best.

Usually unrolling the first iteration should work. In this case I'd unroll the first iteration and then use `chain` and make sure to use combinators in `calculate_root` so that the chain is optimized-away.

BTW in some cases this is the correct unrolling:

```rust
if let Some(first) = iter.next() {
    // maybe do something with first
    for item in iter {
        // process remaining items
    }
}
```

Funnily enough, just today I've found some old code where I do this incorrectly and have to `unwrap` the first item.

**Comment by Kixunil** [comment]
The name still bothers me. I'm thinking that we could just rename `compute_wtxid` to `compute_non_coinbase_wtxid` and not have other methods with this justification:

* It's better to statically know which tx is actually the coinbase (the first in the list) than trying to compute it from transaction structure, therefore other methods are not needed
* The method isn't called often, so the longer name is not too bad. It also discourages use in hashmaps etc, where `compute_ntxid` might be appropriate as it protects against malleability. And if someone is too annoyed by it they can always reexport the method or write an extension trait renaming it.
* The name reasonably clearly indicates that the method is not supposed to be called on coinbase

Alternatively we could rename it to `compute_ncb_wtxid` - it's short and `ncb` will hopefully be sufficient "WTF is that, let's check the doc" trigger.

We could also add a new type called `FullTxid` which is basically what the output of `compute_wtxid` is. Oh, now that I'm thinking about it, maybe we can remove `compute_wtxid` and have `assume_non_coinbase()` on `FullTxid` returning `WTxid`.

**Comment by apoelstra** [comment]
Yeah, that all sounds reasonable. concept ACK `compute_non_coinbase_wtxid`. (I don't like the short name.) Though if you call it on a coinbase tx, what would you prefer to happen?

As for `FullTxid` ... this is sounds like what `ntxid` was supposed to be. Let's keep it in mind when we look at what we want to do with ntxid.

**Comment by Kixunil** [comment]
> what would you prefer to happen?

`debug_assert` sounds good here.

> As for `FullTxid` ... this is sounds like what `ntxid` was supposed to be.

Not really, `ntxid` explicitly *excludes* all signatures (including non-segwit) while `FullTxid` explicitly *inlcudes* all signatures, including witnesses.

**Comment by apoelstra** [comment]
> Not really, `ntxid` explicitly _excludes_ all signatures (including non-segwit) while `FullTxid` explicitly _inlcudes_ all signatures, including witnesses.

Oh, lol, derp you're right. They are pretty-much opposites.
